File: solutions/138_int_break.go
package main

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func integerBreak(n int) int {
	if n < 4 {
		return n - 1
	}
	dp := make([]int, n+1)
	dp[2] = 1
	for i := 4; i <= n; i++ {
		dp[i] = max(
			max(2*(i-2), 2*dp[i-2]),
			max(3*(i-3), 3*dp[i-3]),
		)
	}
	return dp[n]
}

File: solutions/__138_int_break.cpp
//
// Created by peanut on 2021/8/16.
//

#include <bits/stdc++.h>

using namespace std;

int integerBreak(int n) {
    vector<int> dp(n + 1);
    dp[2] = 1;
    for (int i = 3; i <= n; i++) {
        for (int j = 1; j < i - 1; j++) {
            dp[i] = max(dp[i], max((i - j) * j, dp[i - j] * j));
        }
    }
    return dp[n];
}

int main() {
    cout << integerBreak(10) << endl;
    return 0;
}
